File: km/km.java
import java.util.*;

import static java.util.Collections.max;

class KuhnMunkres {
    private Map<String, Float> pole_X = new HashMap<String, Float>();
    private Map<String, Float> pole_Y = new HashMap<String, Float>();
    private Map<String, List<String>> edge_X = new HashMap<String, List<String>>();
    private Map<String, List<String>> edge_Y = new HashMap<String, List<String>>();
    private Map<String, String> matched_Y = new HashMap<String, String>();
    private List<String> marked_X = new ArrayList<String>();
    private List<String> marked_Y = new ArrayList<String>();

    private List<String> list_union(List<String> X, List<String> Y){
        Set<String> hs = new HashSet<String>();
        hs.addAll(X);
        hs.addAll(Y);
        return new ArrayList<String>(hs);
    }

    private List<String> list_intersection(List<String> X, List<String> Y){
        Set<String> hs = new HashSet<String>();
        hs.addAll(X);
        hs.retainAll(Y);
        return new ArrayList<String>(hs);
    }

    private List<String> list_substract(List<String> X, List<String> Y){
        Set<String> hs = new HashSet<String>();
        hs.addAll(X);
        hs.removeAll(Y);
        return new ArrayList<String>(hs);
    }

    private int km_dfs_process(String x, Map<String, List<String>> t_edge_X, Map<String, Map<String, Float>> weights){
        if(!weights.containsKey(x) || !t_edge_X.containsKey(x)) return 0;
        for(String y : t_edge_X.get(x)){
            if(marked_Y.contains(y)) continue;
            if(!matched_Y.containsKey(y)){
                matched_Y.put(y, x);
                marked_Y.add(y);
                return 1;
            }else{
                //try hungarian match
                String t = matched_Y.get(y);
                matched_Y.put(y, x);
                marked_Y.add(y);
                int res = km_dfs_process(t, t_edge_X, weights);
                if(res > 0) return 1;
                else matched_Y.put(y, t);
            }
        }
        return 0;
    }

    public float km_dfs(List<String> X, List<String> Y, Map<String, Map<String, Float>> weights, Map<String, String> matched_result){
        pole_X.clear();
        pole_Y.clear();
        edge_X.clear();
        edge_Y.clear();
        matched_Y.clear();
        marked_X.clear();
        marked_Y.clear();

        float matched_weight = 0;
        //pole init
        for(String x : X){
            if(!weights.containsKey(x)) continue;
            pole_X.put(x, max(weights.get(x).values()));
        }
        for(String y : Y){
            pole_Y.put(y, 0F);
        }
        //search
        for(String x : X){
            //cal d
            SortedMap<Float, Map<String, String>> d_arr = new TreeMap<Float, Map<String, String>>();
            for(String i : list_union(marked_X, Arrays.asList(x))){
                for (String j : list_substract(Y, new ArrayList<String>(matched_Y.keySet()))){
                    if (!weights.containsKey(i) || !weights.containsKey(j)) continue;
                    float w = pole_X.get(i) + pole_Y.get(j) - weights.get(i).get(j);
                    if(!d_arr.containsKey(w)) d_arr.put(w, new HashMap<String, String>());
                    d_arr.get(w).put(i, j);
                }
            }
            for(SortedMap.Entry<Float, Map<String, String>> p : d_arr.entrySet()){
                Float d = p.getKey();
                Map<String, Float> t_pole_X = new HashMap<String, Float>();
                Map<String, Float> t_pole_Y = new HashMap<String, Float>();
                t_pole_X.putAll(pole_X);
                t_pole_Y.putAll(pole_Y);
                //adjust pole value
                if(d > 0){
                    for(Map.Entry<String, Float> q : t_pole_X.entrySet()){
                        if(list_union(marked_X, new ArrayList<String>(matched_Y.keySet())).contains(q.getKey())
                                || list_intersection(edge_X.get(q.getKey()), new ArrayList<String>(matched_Y.keySet())).size() > 0){
                            t_pole_X.put(q.getKey(), t_pole_X.get(q.getKey())-d);
                        }
                    }
                    for(Map.Entry<String, Float> q : t_pole_Y.entrySet()){
                        if(matched_Y.containsKey(q.getKey())) t_pole_Y.put(q.getKey(), t_pole_Y.get(q.getKey()) + d);
                    }
                }
                //achieve viability edge
                Map<String, List<String>> t_edge_X = new HashMap<String, List<String>>();
                Map<String, List<String>> t_edge_Y = new HashMap<String, List<String>>();
                for(String i : X){
                    for(String j : Y){
                        if(!weights.containsKey(i) || !weights.get(i).containsKey(j)) continue;
                        if(t_pole_X.get(i) + t_pole_Y.get(j) != weights.get(i).get(j)) continue;
                        if(!t_edge_X.containsKey(i)) t_edge_X.put(i, new ArrayList<String>());
                        if(!t_edge_X.get(i).contains(j)) t_edge_X.get(i).add(j);
                    }
                }
                //do km match
                marked_Y.clear();
                int res = km_dfs_process(x, t_edge_X, weights);
                if(res > 0){
                    marked_X.add(x);
                    edge_X.clear();
                    edge_X.putAll(t_edge_X);
                    pole_X.clear();
                    pole_X.putAll(t_pole_X);
                    pole_Y.clear();
                    pole_Y.putAll(t_pole_Y);
                    break;
                }
            }
        }

        matched_result.clear();
        matched_result.putAll(matched_Y);
        for(Map.Entry<String, String> p : matched_result.entrySet()){
            matched_weight += weights.get(p.getKey()).get(p.getValue());
        }
        return matched_weight;
    }

    public static void main(String[] args){
        Map<String, Map<String, Float>> weights = new HashMap<String, Map<String, Float>>();
        weights.put("x0", new HashMap<String, Float>());
        weights.put("x1", new HashMap<String, Float>());
        weights.put("x2", new HashMap<String, Float>());
        weights.put("x3", new HashMap<String, Float>());
        weights.put("x4", new HashMap<String, Float>());
        weights.put("y0", new HashMap<String, Float>());
        weights.put("y1", new HashMap<String, Float>());
        weights.put("y2", new HashMap<String, Float>());
        weights.put("y3", new HashMap<String, Float>());
        weights.put("y4", new HashMap<String, Float>());
        weights.get("x0").put("y0", 3F);
        weights.get("x0").put("y1", 4F);
        weights.get("x0").put("y2", 6F);
        weights.get("x0").put("y3", 4F);
        weights.get("x0").put("y4", 9F);
        weights.get("x1").put("y0", 6F);
        weights.get("x1").put("y1", 4F);
        weights.get("x1").put("y2", 5F);
        weights.get("x1").put("y3", 3F);
        weights.get("x1").put("y4", 8F);
        weights.get("x2").put("y0", 7F);
        weights.get("x2").put("y1", 5F);
        weights.get("x2").put("y2", 3F);
        weights.get("x2").put("y3", 4F);
        weights.get("x2").put("y4", 2F);
        weights.get("x3").put("y0", 6F);
        weights.get("x3").put("y1", 3F);
        weights.get("x3").put("y2", 2F);
        weights.get("x3").put("y3", 2F);
        weights.get("x3").put("y4", 5F);
        weights.get("x4").put("y0", 8F);
        weights.get("x4").put("y1", 4F);
        weights.get("x4").put("y2", 5F);
        weights.get("x4").put("y3", 4F);
        weights.get("x4").put("y4", 7F);
        weights.get("y0").put("x0", 3F);
        weights.get("y0").put("x1", 6F);
        weights.get("y0").put("x2", 7F);
        weights.get("y0").put("x3", 6F);
        weights.get("y0").put("x4", 8F);
        weights.get("y1").put("x0", 4F);
        weights.get("y1").put("x1", 4F);
        weights.get("y1").put("x2", 5F);
        weights.get("y1").put("x3", 3F);
        weights.get("y1").put("x4", 4F);
        weights.get("y2").put("x0", 6F);
        weights.get("y2").put("x1", 5F);
        weights.get("y2").put("x2", 3F);
        weights.get("y2").put("x3", 2F);
        weights.get("y2").put("x4", 5F);
        weights.get("y3").put("x0", 4F);
        weights.get("y3").put("x1", 3F);
        weights.get("y3").put("x2", 4F);
        weights.get("y3").put("x3", 2F);
        weights.get("y3").put("x4", 4F);
        weights.get("y4").put("x0", 9F);
        weights.get("y4").put("x1", 8F);
        weights.get("y4").put("x2", 2F);
        weights.get("y4").put("x3", 5F);
        weights.get("y4").put("x4", 7F);
        List<String> X = new ArrayList<String>();
        List<String> Y = new ArrayList<String>();
        X.add("x0");
        X.add("x1");
        X.add("x2");
        X.add("x3");
        X.add("x4");
        Y.add("y0");
        Y.add("y1");
        Y.add("y2");
        Y.add("y3");
        Y.add("y4");
        KuhnMunkres km = new KuhnMunkres();
        Map<String, String> matched_result = new HashMap<String, String>();
        float matched_weight = km.km_dfs(X, Y, weights, matched_result);
        System.out.println(matched_weight);
        for (Map.Entry<String, String> p : matched_result.entrySet()){
            System.out.println(p.getKey() + "--" + p.getValue());
        }
    }
}

File: km/km.py
import copy

class KuhnMunkres:
    def __init__(self, X, Y, weights):
        self.X = X
        self.Y = Y
        self.weights = weights

    #km algorithm with dfs search
    #https://www.cnblogs.com/zpfbuaa/p/7218607.html
    #input:nodes_X,nodes_Y,weights vector
    #output:1.matched_num, 2.matched_weight 3.matched_map
    def km_dfs(self):
        global pole_X,pole_Y,edge_X,matched_Y,marked_X,marked_Y
        pole_X = {}
        pole_Y = {}
        matched_Y = {}
        marked_X = []
        #pole init
        for x in X:
            if x not in weights: continue
            pole_X[x] = max(weights[x].values())
        for y in Y:
            pole_Y[y] = 0
        #search
        for x in X:
            #cal d
            d_arr = [(i,j,(pole_X[i] + pole_Y[j]) - weights[i][j]) \
                     for i in set(marked_X) | set([x]) \
                     for j in set(Y)-set(matched_Y.keys()) \
                     if i in weights and j in weights[i]]
            sorted_d_arr = sorted(d_arr, key=lambda d:d[2])
            for d in set([s[2] for s in sorted_d_arr]):
                t_pole_X = copy.deepcopy(pole_X)
                t_pole_Y = copy.deepcopy(pole_Y)
                #adjust pole value
                if d > 0:
                    for k,v in t_pole_X.items():
                        if (k in set(marked_X) | set([x])) \
                                or len(set(edge_X[k]) & set(matched_Y.keys())) > 0:
                            t_pole_X[k] -= d
                    for k,v in t_pole_Y.items():
                        if k in matched_Y:
                            t_pole_Y[k] += d
                #achieve viability edge
                t_edge_X = {}
                for i in X:
                    for j in Y:
                        if i not in weights or j not in weights[i]: continue
                        if t_pole_X[i] + t_pole_Y[j] == weights[i][j]:
                            if i not in t_edge_X:
                                t_edge_X[i] = []
                            if j not in t_edge_X[i]:
                                t_edge_X[i].append(j)
                #do km match
                marked_Y = []
                def km_dfs_process(x, t_edge_X):
                    if x not in weights or x not in t_edge_X:
                        return 0
                    for y in t_edge_X[x]:
                        if y in marked_Y:
                            continue
                        if y not in matched_Y:
                            matched_Y[y] = x
                            marked_Y.append(y)
                            return 1
                        else:
                            # try hungarian match
                            t = matched_Y[y]
                            matched_Y[y] = x
                            marked_Y.append(y)
                            res = km_dfs_process(t, t_edge_X)
                            if res > 0:
                                return 1
                            else:
                                matched_Y[y] = t
                    return 0
                res = km_dfs_process(x, t_edge_X)
                if res > 0:
                    marked_X.append(x)
                    edge_X = copy.deepcopy(t_edge_X)
                    pole_X = copy.deepcopy(t_pole_X)
                    pole_Y = copy.deepcopy(t_pole_Y)
                    break
        return len(matched_Y), sum([weights[j][i] for i,j in matched_Y.items()]), matched_Y


weights = {}
weights['x0'] = {'y0':3, 'y1':4, 'y2':6, 'y3':4, 'y4':9}
weights['x1'] = {'y0':6, 'y1':4, 'y2':5, 'y3':3, 'y4':8}
weights['x2'] = {'y0':7, 'y1':5, 'y2':3, 'y3':4, 'y4':2}
weights['x3'] = {'y0':6, 'y1':3, 'y2':2, 'y3':2, 'y4':5}
weights['x4'] = {'y0':8, 'y1':4, 'y2':5, 'y3':4, 'y4':7}
weights['y0'] = {'x0':3, 'x1':6, 'x2':7, 'x3':6, 'x4':8}
weights['y1'] = {'x0':4, 'x1':4, 'x2':5, 'x3':3, 'x4':4}
weights['y2'] = {'x0':6, 'x1':5, 'x2':3, 'x3':2, 'x4':5}
weights['y3'] = {'x0':4, 'x1':3, 'x2':4, 'x3':2, 'x4':4}
weights['y4'] = {'x0':9, 'x1':8, 'x2':2, 'x3':5, 'x4':7}
X = ['x0', 'x1', 'x2', 'x3', 'x4']
Y = ['y0', 'y1', 'y2', 'y3', 'y4']
km = KuhnMunkres(X, Y, weights)
print km.km_dfs()
